C++11 compile failure

When building hokuyoaist with a C++11 compiler, there are a lot of errors of the following variety:

```
/home/rich/fedora/hokuyoaist/hokuyoaist-b87a88aa66853d3c9d901d4e6be729c5fe69aae0/src/hokuyo_errors.cpp: In member function ‘virtual const char* hokuyoaist::BaseError::what() const’:
/home/rich/fedora/hokuyoaist/hokuyoaist-b87a88aa66853d3c9d901d4e6be729c5fe69aae0/src/hokuyo_errors.cpp:386:8: error: no match for ‘operator<<’ (operand types are ‘const stringstream {aka const std::__cxx11::basic_stringstream<char>}’ and ‘std::__cxx11::string {aka std::__cxx11::basic_string<char>}’)
     ss << std::string(error_type_, sizeof(error_type_)) << " (" << desc_code_ << "): " <<
     ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6.1.1/istream:39:0,
                 from /usr/include/c++/6.1.1/sstream:38,
                 from /home/rich/fedora/hokuyoaist/hokuyoaist-b87a88aa66853d3c9d901d4e6be729c5fe69aae0/include/hokuyoaist/hokuyo_errors.h:32,
                 from /home/rich/fedora/hokuyoaist/hokuyoaist-b87a88aa66853d3c9d901d4e6be729c5fe69aae0/src/hokuyo_errors.cpp:29:
/usr/include/c++/6.1.1/ostream:108:7: note: candidate: std::basic_ostream<_CharT, _Traits>::__ostream_type& std::basic_ostream<_CharT, _Traits>::operator<<(std::basic_ostream<_CharT, _Traits>::__ostream_type& (*)(std::basic_ostream<_CharT, _Traits>::__ostream_type&)) [with _CharT = char; _Traits = std::char_traits<char>; std::basic_ostream<_CharT, _Traits>::__ostream_type = std::basic_ostream<char>]
       operator<<(__ostream_type& (*__pf)(__ostream_type&))
       ^~~~~~~~
/usr/include/c++/6.1.1/ostream:108:7: note:   no known conversion for argument 1 from ‘std::__cxx11::string {aka std::__cxx11::basic_string<char>}’ to ‘std::basic_ostream<char>::__ostream_type& (*)(std::basic_ostream<char>::__ostream_type&) {aka std::basic_ostream<char>& (*)(std::basic_ostream<char>&)}’
/usr/include/c++/6.1.1/ostream:117:7: note: candidate: std::basic_ostream<_CharT, _Traits>::__ostream_type& std::basic_ostream<_CharT, _Traits>::operator<<(std::basic_ostream<_CharT, _Traits>::__ios_type& (*)(std::basic_ostream<_CharT, _Traits>::__ios_type&)) [with _CharT = char; _Traits = std::char_traits<char>; std::basic_ostream<_CharT, _Traits>::__ostream_type = std::basic_ostream<char>; std::basic_ostream<_CharT, _Traits>::__ios_type = std::basic_ios<char>]
       operator<<(__ios_type& (*__pf)(__ios_type&))
       ^~~~~~~~
/usr/include/c++/6.1.1/ostream:117:7: note:   no known conversion for argument 1 from ‘std::__cxx11::string {aka std::__cxx11::basic_string<char>}’ to ‘std::basic_ostream<char>::__ios_type& (*)(std::basic_ostream<char>::__ios_type&) {aka std::basic_ios<char>& (*)(std::basic_ios<char>&)}’
/usr/include/c++/6.1.1/ostream:127:7: note: candidate: std::basic_ostream<_CharT, _Traits>::__ostream_type& std::basic_ostream<_CharT, _Traits>::operator<<(std::ios_base& (*)(std::ios_base&)) [with _CharT = char; _Traits = std::char_traits<char>; std::basic_ostream<_CharT, _Traits>::__ostream_type = std::basic_ostream<char>]
       operator<<(ios_base& (*__pf) (ios_base&))
       ^~~~~~~~
/usr/include/c++/6.1.1/ostream:127:7: note:   no known conversion for argument 1 from ‘std::__cxx11::string {aka std::__cxx11::basic_string<char>}’ to ‘std::ios_base& (*)(std::ios_base&)’
/usr/include/c++/6.1.1/ostream:166:7: note: candidate: std::basic_ostream<_CharT, _Traits>::__ostream_type& std::basic_ostream<_CharT, _Traits>::operator<<(long int) [with _CharT = char; _Traits = std::char_traits<char>; std::basic_ostream<_CharT, _Traits>::__ostream_type = std::basic_ostream<char>]
       operator<<(long __n)
       ^~~~~~~~
/usr/include/c++/6.1.1/ostream:166:7: note:   no known conversion for argument 1 from ‘std::__cxx11::string {aka std::__cxx11::basic_string<char>}’ to ‘long int’
/usr/include/c++/6.1.1/ostream:170:7: note: candidate: std::basic_ostream<_CharT, _Traits>::__ostream_type& std::basic_ostream<_CharT, _Traits>::operator<<(long unsigned int) [with _CharT = char; _Traits = std::char_traits<char>; std::basic_ostream<_CharT, _Traits>::__ostream_type = std::basic_ostream<char>]
       operator<<(unsigned long __n)
       ^~~~~~~~
/usr/include/c++/6.1.1/ostream:170:7: note:   no known conversion for argument 1 from ‘std::__cxx11::string {aka std::__cxx11::basic_string<char>}’ to ‘long unsigned int’
/usr/include/c++/6.1.1/ostream:174:7: note: candidate: std::basic_ostream<_CharT, _Traits>::__ostream_type& std::basic_ostream<_CharT, _Traits>::operator<<(bool) [with _CharT = char; _Traits = std::char_traits<char>; std::basic_ostream<_CharT, _Traits>::__ostream_type = std::basic_ostream<char>]
       operator<<(bool __n)
       ^~~~~~~~
...
```

I think the root cause is that issue #6 changed the what() function prototype for the exceptions in hokuyo_exception to be a const function as required by c++11 std::exception::what().  However, the function body is not const, as each of the what() functions modify the member `std::stringstream ss` variable.

To fix this, it seems like the error description that is being built in `what()` should be built in the exception constructor, and stored in `ss_str`.  Then the `what()` function can just return `ss_str.c_str()`.  Doing that could allow you to remove the stringstream `ss` from the class entirely in favor of a temporary stringstream in each constructor.

--- a/include/hokuyoaist/hokuyo_errors.h
+++ b/include/hokuyoaist/hokuyo_errors.h
@@ -72,7 +72,7 @@ class HOKUYOAIST_EXPORT BaseError : public std::exception
         */
         BaseError(unsigned int desc_code, char const* error_type);
         BaseError(BaseError const& rhs);
-        virtual ~BaseError() throw() {};
+        virtual ~BaseError() throw() {}
 
         virtual unsigned int desc_code() const throw()
             { return desc_code_; }
@@ -80,19 +80,15 @@ class HOKUYOAIST_EXPORT BaseError : public std::exception
         virtual char const* error_type() const throw()
             { return error_type_; }
 
-#if __cplusplus >= 201103L
         virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** Description code for use with the error string table. */
         unsigned int desc_code_;
 
         /** Formatted description of the error. */
-        std::stringstream ss;
-        std::string ss_str;
+        std::string error_str_;
+
         /** String representation of the error. */
         char error_type_[32];
 }; //class BaseError
@@ -128,7 +124,7 @@ class HOKUYOAIST_EXPORT RuntimeError : public BaseError
         RuntimeError(unsigned int desc_code, char const* error_type)
             : BaseError(desc_code, error_type)
         {}
-        virtual ~RuntimeError() throw() {};
+        virtual ~RuntimeError() throw() {}
 }; // class RuntimeError
 
 
@@ -165,21 +161,11 @@ class HOKUYOAIST_EXPORT BaudrateError: public RuntimeError
         /** @brief Baud rate error constructor.
 
         @param baud The bad baud rate. */
-        BaudrateError(unsigned int baud)
-            : RuntimeError(6, "BaudrateError"), baud_(baud)
-        {}
-        BaudrateError(BaudrateError const& rhs)
-            : RuntimeError(rhs), baud_(rhs.baud())
-        {}
+        BaudrateError(unsigned int baud);
+        BaudrateError(BaudrateError const& rhs);
 
         unsigned int baud() const throw()
             { return baud_; }
-
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** Baud rate that caused the error. */
@@ -362,26 +348,14 @@ class HOKUYOAIST_EXPORT ChecksumError: public ProtocolError
 
         @param expected The expected checksum.
         @param calculated The calculated checksum. */
-        ChecksumError(int expected, int calculated)
-            : ProtocolError(24, "ChecksumError"), expected_(expected),
-            calculated_(calculated)
-        {}
-        ChecksumError(ChecksumError const& rhs)
-            : ProtocolError(rhs), expected_(rhs.expected()),
-            calculated_(rhs.calculated())
-        {}
+        ChecksumError(int expected, int calculated);
+        ChecksumError(ChecksumError const& rhs);
 
         virtual int expected() const throw()
             { return expected_; }
 
         virtual int calculated() const throw()
             { return calculated_; }
-
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** Expected checksum value. */
@@ -424,12 +398,6 @@ class HOKUYOAIST_EXPORT UnknownLineError: public ProtocolError
         virtual char const* const line() const throw()
             { return line_; }
 
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
-
     protected:
         /** The mystery line. */
         char line_[128];
@@ -452,12 +420,6 @@ class HOKUYOAIST_EXPORT ParseError: public ProtocolError
 
         virtual char const* const type() const throw()
             { return type_; }
-
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** The bad line. */
@@ -485,20 +447,8 @@ class HOKUYOAIST_EXPORT ResponseError: public ProtocolError
 
         @param error The two-byte error code received.
         @param cmd The command that caused the error. */
-        ResponseError(char const* const error, char const* const cmd)
-            : ProtocolError(30, "ResponseError")
-        {
-            error_[0] = error[0]; error_[1] = error[1];
-            cmd_[0] = cmd[0]; cmd_[1] = cmd[1];
-        }
-        ResponseError(ResponseError const& rhs)
-            : ProtocolError(rhs)
-        {
-            error_[0] = rhs.error_code()[0];
-            error_[1] = rhs.error_code()[1];
-            cmd_[0] = rhs.cmd_code()[0];
-            cmd_[1] = rhs.cmd_code()[1];
-        }
+        ResponseError(char const* const error, char const* const cmd);
+        ResponseError(ResponseError const& rhs);
 
         /// Get the two-byte error code as a non-null-terminated array.
         virtual char const* const error_code() const throw()
@@ -507,12 +457,6 @@ class HOKUYOAIST_EXPORT ResponseError: public ProtocolError
         /// Get the two-byte command code as a non-null-terminated array.
         virtual char const* const cmd_code() const throw()
             { return cmd_; }
-
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** Error code as defined in SCIP2 (two bytes). */
@@ -530,15 +474,8 @@ class HOKUYOAIST_EXPORT Scip1ResponseError: public ProtocolError
 
         @param error The two-byte error code received.
         @param cmd The command that caused the error. */
-        Scip1ResponseError(char error, char cmd)
-            : ProtocolError(30, "Scip1ResponseError"),
-            error_(error), cmd_(cmd)
-        {}
-        Scip1ResponseError(Scip1ResponseError const& rhs)
-            : ProtocolError(rhs), error_(rhs.error_code()),
-            cmd_(rhs.cmd_code())
-        {}
-
+        Scip1ResponseError(char error, char cmd);
+        Scip1ResponseError(Scip1ResponseError const& rhs);
         /// Get the one-byte error code.
         virtual char error_code() const throw()
             { return error_; }
@@ -546,12 +483,6 @@ class HOKUYOAIST_EXPORT Scip1ResponseError: public ProtocolError
         /// Get the one-byte command code.
         virtual char cmd_code() const throw()
             { return cmd_; }
-
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** Error code as defined in SCIP2 (two bytes). */
@@ -569,20 +500,8 @@ class HOKUYOAIST_EXPORT CommandEchoError: public ProtocolError
 
         @param cmd The two-byte command code expected.
         @param echo The two-byte command echo received. */
-        CommandEchoError(char const* const cmd, char const* const echo)
-            : ProtocolError(31, "CommandEchoError")
-        {
-            cmd_[0] = cmd[0]; cmd_[1] = cmd[1];
-            echo_[0] = echo[0]; echo_[1] = echo[1];
-        }
-        CommandEchoError(CommandEchoError const& rhs)
-            : ProtocolError(rhs)
-        {
-            cmd_[0] = rhs.cmd_code()[0];
-            cmd_[1] = rhs.cmd_code()[1];
-            echo_[0] = rhs.cmd_echo()[0];
-            echo_[1] = rhs.cmd_echo()[1];
-        }
+        CommandEchoError(char const* const cmd, char const* const echo);
+        CommandEchoError(CommandEchoError const& rhs);
 
         /// Get the two-byte command code as a non-null-terminated array.
         virtual char const* const cmd_code() const throw()
@@ -591,12 +510,6 @@ class HOKUYOAIST_EXPORT CommandEchoError: public ProtocolError
         /// Get the two-byte command echo as a non-null-terminated array.
         virtual char const* const cmd_echo() const throw()
             { return echo_; }
-
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** Command that triggered the error, from SCIP2 (two bytes). */
@@ -613,27 +526,12 @@ class HOKUYOAIST_EXPORT ParamEchoError: public ProtocolError
         /** @brief Parameter echo error constructor.
 
         @param cmd The two-byte command code sent. */
-        ParamEchoError(char const* const cmd)
-            : ProtocolError(32, "ParamEchoError")
-        {
-            cmd_[0] = cmd[0]; cmd_[1] = cmd[1];
-        }
-        ParamEchoError(ParamEchoError const& rhs)
-            : ProtocolError(rhs)
-        {
-            cmd_[0] = rhs.cmd_code()[0];
-            cmd_[1] = rhs.cmd_code()[1];
-        }
+        ParamEchoError(char const* const cmd);
+        ParamEchoError(ParamEchoError const& rhs);
 
         /// Get the two-byte command code as a non-null-terminated array.
         virtual char const* const cmd_code() const throw()
             { return cmd_; }
-
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** Command that triggered the error, from SCIP2 (two bytes). */
@@ -649,26 +547,14 @@ class HOKUYOAIST_EXPORT InsufficientBytesError: public ProtocolError
 
         @param num The number of bytes received.
         @param line_length The length of the line. */
-        InsufficientBytesError(int num, int line_length)
-            : ProtocolError(33, "InsufficientBytesError"),
-            num_(num), line_length_(line_length)
-        {}
-        InsufficientBytesError(InsufficientBytesError const& rhs)
-            : ProtocolError(rhs), num_(rhs.num()),
-            line_length_(rhs.line_length())
-        {}
+        InsufficientBytesError(int num, int line_length);
+        InsufficientBytesError(InsufficientBytesError const& rhs);
 
         virtual int num() const throw()
             { return num_; }
 
         virtual int line_length() const throw()
             { return line_length_; }
-
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** Number of bytes available. */
@@ -686,26 +572,14 @@ class HOKUYOAIST_EXPORT LineLengthError: public ProtocolError
 
         @param length The number of bytes received.
         @param expected The expected length of the line. */
-        LineLengthError(int length, int expected)
-            : ProtocolError(34, "LineLengthError"),
-            length_(length), expected_(expected)
-        {}
-        LineLengthError(LineLengthError const& rhs)
-            : ProtocolError(rhs), length_(rhs.length()),
-            expected_(rhs.expected())
-        {}
+        LineLengthError(int length, int expected);
+        LineLengthError(LineLengthError const& rhs);
 
         virtual int length() const throw()
             { return length_; }
 
         virtual int expected() const throw()
             { return expected_; }
-
-#if __cplusplus >= 201103L
-        virtual const char* what() const throw();
-#else
-        virtual const char* what() throw();
-#endif
 
     protected:
         /** The received line length. */
@@ -714,7 +588,7 @@ class HOKUYOAIST_EXPORT LineLengthError: public ProtocolError
         int expected_;
 }; // class LineLengthError
 
-}; // namespace hokuyoaist
+} // namespace hokuyoaist
 
 /** @} */
 

--- a/python/hokuyo_aist.cpp
+++ b/python/hokuyo_aist.cpp
@@ -43,8 +43,7 @@ class BaseErrorWrap
         {
             return BaseError::error_type();
         }
-
-        const char* what() throw()
+        const char* what() const throw()
         {
             if (boost::python::override f = get_override("what"))
             {
@@ -52,7 +51,7 @@ class BaseErrorWrap
             }
             return BaseError::what();
         }
-        const char* default_what() throw()
+        const char* default_what() const throw()
         {
             return BaseError::what();
         }

--- a/src/hokuyo_errors.cpp
+++ b/src/hokuyo_errors.cpp
@@ -367,59 +367,46 @@ BaseError::BaseError(unsigned int desc_code, char const* error_type)
     : desc_code_(desc_code)
 {
     strncpy(error_type_, error_type, 32);
+    std::stringstream ss;
+    ss << error_type_ << " (" << desc_code_ << "): " <<
+        desc_code_to_string(desc_code_);
+    error_str_ = ss.str();
 }
 
 
 BaseError::BaseError(BaseError const& rhs)
-    : desc_code_(rhs.desc_code())
+    : desc_code_(rhs.desc_code()),
+    error_str_(rhs.error_str_)
 {
     strncpy(error_type_, rhs.error_type(), 32);
 }
 
-
-#if __cplusplus >= 201103L
 const char* BaseError::what() const throw()
-#else
-const char* BaseError::what() throw()
-#endif
 {
-    ss << error_type_ << " (" << desc_code_ << "): " <<
-        desc_code_to_string(desc_code_);
-    ss_str = ss.str();
-    return ss_str.c_str();;
+    return error_str_.c_str();
 }
 
-
-#if __cplusplus >= 201103L
-const char* BaudrateError::what() const throw()
-#else
-const char* BaudrateError::what() throw()
-#endif
+BaudrateError::BaudrateError(unsigned int baud)
+    : RuntimeError(6, "BaudrateError"), baud_(baud)
 {
-    RuntimeError::what();
+    std::stringstream ss;
     ss << baud_;
-    ss_str = ss.str();
-    return ss_str.c_str();;
+    error_str_ += ss.str();
 }
 
-
-#if __cplusplus >= 201103L
-const char* ChecksumError::what() const throw()
-#else
-const char* ChecksumError::what() throw()
-#endif
+ChecksumError::ChecksumError(int expected, int calculated)
+    : ProtocolError(24, "ChecksumError"), expected_(expected), calculated_(calculated)
 {
-    ProtocolError::what();
+    std::stringstream ss;
     ss << "expected " << expected_ << ", calculated " << calculated_;
-    ss_str = ss.str();
-    return ss_str.c_str();;
+    error_str_ += ss.str();
 }
-
 
 UnknownLineError::UnknownLineError(char const* const line)
     : ProtocolError(27, "UnknownLineError")
 {
     strncpy(line_, line, 128);
+    error_str_ += std::string(line);
 }
 
 
@@ -429,25 +416,14 @@ UnknownLineError::UnknownLineError(UnknownLineError const& rhs)
     strncpy(line_, rhs.line(), 128);
 }
 
-
-#if __cplusplus >= 201103L
-const char* UnknownLineError::what() const throw()
-#else
-const char* UnknownLineError::what() throw()
-#endif
-{
-    ProtocolError::what();
-    ss << line_;
-    ss_str = ss.str();
-    return ss_str.c_str();;
-}
-
-
 ParseError::ParseError(char const* const line, char const* const type)
     : ProtocolError(28, "ParseError")
 {
     strncpy(line_, line, 128);
     strncpy(type_, type, 16);
+    std::stringstream ss;
+    ss << "Line type: " << type_ << ". Line: " << line_;
+    error_str_ += ss.str();
 }
 
 
@@ -458,102 +434,107 @@ ParseError::ParseError(ParseError const& rhs)
     strncpy(type_, rhs.type(), 16);
 }
 
-
-#if __cplusplus >= 201103L
-const char* ParseError::what() const throw()
-#else
-const char* ParseError::what() throw()
-#endif
+ResponseError::ResponseError(char const* const error, char const* const cmd)
+    : ProtocolError(30, "ResponseError")
 {
-    ProtocolError::what();
-    ss << "Line type: " << type_ << ". Line: " << line_;
-    ss_str = ss.str();
-    return ss_str.c_str();;
-}
-
-
-#if __cplusplus >= 201103L
-const char* ResponseError::what() const throw()
-#else
-const char* ResponseError::what() throw()
-#endif
-{
-    ProtocolError::what();
+    error_[0] = error[0]; error_[1] = error[1];
+    cmd_[0] = cmd[0]; cmd_[1] = cmd[1];
+    std::stringstream ss;
     ss << " Command: " << cmd_[0] << cmd_[1];
     ss << " Error : (" << error_[0] << error_[1] << ") " <<
         scip2_error_to_string(error_, cmd_);
-    ss_str = ss.str();
-    return ss_str.c_str();;
+    error_str_ += ss.str();
+}
+ResponseError::ResponseError(ResponseError const& rhs)
+    : ProtocolError(rhs)
+{
+    error_[0] = rhs.error_code()[0];
+    error_[1] = rhs.error_code()[1];
+    cmd_[0] = rhs.cmd_code()[0];
+    cmd_[1] = rhs.cmd_code()[1];
 }
 
-
-#if __cplusplus >= 201103L
-const char* Scip1ResponseError::what() const throw()
-#else
-const char* Scip1ResponseError::what() throw()
-#endif
+Scip1ResponseError::Scip1ResponseError(char error, char cmd)
+    : ProtocolError(30, "Scip1ResponseError"),
+    error_(error), cmd_(cmd)
 {
-    ProtocolError::what();
+    std::stringstream ss;
     ss << " Command: " << cmd_;
     ss << " Error : " << error_;
-    ss_str = ss.str();
-    return ss_str.c_str();;
+    error_str_ += ss.str();
+}
+Scip1ResponseError::Scip1ResponseError(Scip1ResponseError const& rhs)
+    : ProtocolError(rhs), error_(rhs.error_code()),
+    cmd_(rhs.cmd_code())
+{
 }
 
-
-#if __cplusplus >= 201103L
-const char* CommandEchoError::what() const throw()
-#else
-const char* CommandEchoError::what() throw()
-#endif
+CommandEchoError::CommandEchoError(char const* const cmd, char const* const echo)
+    : ProtocolError(31, "CommandEchoError")
 {
-    ProtocolError::what();
+    cmd_[0] = cmd[0]; cmd_[1] = cmd[1];
+    echo_[0] = echo[0]; echo_[1] = echo[1];
+    std::stringstream ss;
     ss << " Command: " << cmd_[0] << cmd_[1];
     ss << " Received echo: " << echo_[0] << echo_[1];
-    ss_str = ss.str();
-    return ss_str.c_str();;
+    error_str_ += ss.str();
+
+}
+CommandEchoError::CommandEchoError(CommandEchoError const& rhs)
+    : ProtocolError(rhs)
+{
+    cmd_[0] = rhs.cmd_code()[0];
+    cmd_[1] = rhs.cmd_code()[1];
+    echo_[0] = rhs.cmd_echo()[0];
+    echo_[1] = rhs.cmd_echo()[1];
 }
 
-
-#if __cplusplus >= 201103L
-const char* ParamEchoError::what() const throw()
-#else
-const char* ParamEchoError::what() throw()
-#endif
+ParamEchoError::ParamEchoError(char const* const cmd)
+    : ProtocolError(32, "ParamEchoError")
 {
-    ProtocolError::what();
+    cmd_[0] = cmd[0]; cmd_[1] = cmd[1];
+    std::stringstream ss;
     ss << " Command: " << cmd_[0] << cmd_[1];
-    ss_str = ss.str();
-    return ss_str.c_str();;
+    error_str_ += ss.str();
 }
 
-
-#if __cplusplus >= 201103L
-const char* InsufficientBytesError::what() const throw()
-#else
-const char* InsufficientBytesError::what() throw()
-#endif
+ParamEchoError::ParamEchoError(ParamEchoError const& rhs)
+    : ProtocolError(rhs)
 {
-    ProtocolError::what();
+    cmd_[0] = rhs.cmd_code()[0];
+    cmd_[1] = rhs.cmd_code()[1];
+}
+
+InsufficientBytesError::InsufficientBytesError(int num, int line_length)
+    : ProtocolError(33, "InsufficientBytesError"),
+    num_(num), line_length_(line_length)
+{
+    std::stringstream ss;
     ss << " Number of bytes: " << num_;
     ss << " Line length: " << line_length_;
-    ss_str = ss.str();
-    return ss_str.c_str();;
+    error_str_ += ss.str();
+}
+InsufficientBytesError::InsufficientBytesError(InsufficientBytesError const& rhs)
+    : ProtocolError(rhs), num_(rhs.num()),
+    line_length_(rhs.line_length())
+{
 }
 
-
-#if __cplusplus >= 201103L
-const char* LineLengthError::what() const throw()
-#else
-const char* LineLengthError::what() throw()
-#endif
+LineLengthError::LineLengthError(int length, int expected)
+    : ProtocolError(34, "LineLengthError"),
+    length_(length), expected_(expected)
 {
-    ProtocolError::what();
+    std::stringstream ss;
     ss << " Received length: " << length_;
     ss << " Expected line length: " << expected_;
-    ss_str = ss.str();
-    return ss_str.c_str();;
+    error_str_ += ss.str();
 }
 
-}; // namespace hokuyoaist
+LineLengthError::LineLengthError(LineLengthError const& rhs)
+    : ProtocolError(rhs), length_(rhs.length()),
+    expected_(rhs.expected())
+{
+}
+
+} // namespace hokuyoaist
 

